src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -265,75 +264,67 @@ public static bool FilterPackages(IEnumerable<FrameworkPackages> packages, ListP
         }
 
         /// <summary>
-        /// Fetches the latest versions for all of the packages that are
-        /// to be listed
+        /// Get package metadata from all sources
         /// </summary>
-        /// <param name="packages">The packages found in a project</param>
-        /// <param name="listPackageArgs">List args for the token and source provider</param>
-        /// <returns>A data structure like packages, but includes the latest versions</returns>
-        private async Task AddLatestVersionsAsync(
-            IEnumerable<FrameworkPackages> packages,
+        /// <param name="targetFrameworks">A <see cref="FrameworkPackages"/> per project target framework</param>
+        /// <param name="listPackageArgs">List command args</param>
+        /// <returns>A dictionary where the key is the package id, and the value is a list of <see cref="IPackageSearchMetadata"/>.</returns>
+        private async Task<Dictionary<string, List<IPackageSearchMetadata>>> GetPackageMetadataAsync(
+            List<FrameworkPackages> targetFrameworks,
             ListPackageArgs listPackageArgs)
         {
-            //Unique Dictionary for packages and list of latest versions to handle different sources
-            var packagesVersionsDict = new Dictionary<string, IList<IPackageSearchMetadata>>();
-            AddPackagesToDict(packages, packagesVersionsDict);
+            List<string> allPackages = GetAllPackageIdentifiers(targetFrameworks, listPackageArgs.IncludeTransitive);
+            var packageMetadataById = new Dictionary<string, List<IPackageSearchMetadata>>(capacity: allPackages.Count);
 
-            //Prepare requests for each of the packages
-            var getLatestVersionsRequests = new List<Task>();
-            foreach (var package in packagesVersionsDict)
+            // We're (probably) going to make a bunch of HTTP requests, so limit to max 4 packages in parallel to avoid being unkind to servers.
+            // Note, each package will also run all sources in parallel, so the max concurrent HTTP requests is higher.
+            int taskCount = Math.Min(allPackages.Count, 4);
```

Comment:
Should this `4` would be defined as const on top?

Reply:
why? It's only used in one place, and there's a comment explaining it.  Constants are useful when it's used in multiple places, so when the value changes, it doesn't need to be changed in multiple places. But that doesn't apply here.